src/window.rs

```diff
@@ -60,16 +58,46 @@ impl Window {
             fb
         };
 
-        if !self.crtc_set {
-            self.crtc_set = true;
+        if self.frame_count == 0 {
+            // NOTE(mbernat): It's possible to avoid initial mode setting
+            // since we're keeping the previous mode: we could just call page_flip directly.
             self.drm_display.set_mode_with_framebuffer(Some(fb));
         } else {
             self.drm_display.page_flip(fb);
         }
+
+        // NOTE(mbernat): This buffer object returns to the surface's queue upon destruction.
+        // If we were to release it here, it would be again available from
+        // `Surface::lock_front_buffer()` next and the app would be effectively single-buffered.
+        // So, we keep it around for one frame, which is fine for double buffering.
+        self.previous_bo.replace(buffer_object);
+
+        self.frame_count += 1;
     }
 
     pub fn restore_original_display(&self) {
-        self.drm_display.set_mode_with_framebuffer(self.drm_display.crtc.framebuffer());
+        let handle = self.drm_display.crtc.framebuffer().unwrap();
+        if self.drm_display.gbm_device.get_framebuffer(handle).is_ok() {
+            self.drm_display.set_mode_with_framebuffer(self.drm_display.crtc.framebuffer());
+        }
+    }
+
+    pub fn draw(&mut self, mut drawer: impl FnMut()) {
+        drawer();
+        // SAFETY: eglSwapBuffers is called by `frame.finish()` in drawer()
```

Comment:
We can't _really_ be sure `eglSwapBuffers()` is called in the closure since this is a public function, so the SAFETY comment is a bit misleading. Are there any true unsafe consequences of calling `self.present` if `eglSwapBuffers()` has not been called?

I'm hoping in the worst case, just nothing is drawn on the screen. We can probably start by marking `self.present()` as safe, with more granular unsafe blocks within.

Reply:
Yeah, this thing became unsound again after code shuffling. We need to call `frame.finish()` ourselves with a separate abstraction. I wonder whether `glium` already has some affordance for this as this frame API is very easy to misuse.

I'm not sure what the consequences are since this has to do again with the poorly documented GBM surface. My guess would also be just broken presentation but I can also imagine that the very first time you request a buffer from the surface it contains uninitialized memory and needs to be written to at least once (e.g. by `eglSwapBuffers` that implicitly calls `glFlush`: https://registry.khronos.org/EGL/sdk/docs/man/html/eglSwapBuffers.xhtml).

```
    /// # Safety
    /// This function must be called exactly once after calling
    /// `eglSwapBuffers`.  Calling it before any `eglSwapBuffers` has happened
    /// on the surface or two or more times after `eglSwapBuffers` is an
    /// error and may cause undefined behavior.
```

OTOH, all of this seems really fishy, GBM should not care whether we're using EGL at all. I guess the comment just reflects the most common way this stuff is being used, rather than the root safety requirements.